Добавил возможность регистрации пользователя

coder1coder opened the pull request:
Task: #40 

+ изменена модель User, добавлена Role
+ добавлен класс, интерфейс репозитория пользователя
+ добавлен класс сервиса для работы с пользователем
+ добавлен метод регистрации в контроллер пользователей

// Непонятно кто регулирует назначение роли пользователя, поэтому DTO при регистрации не содержит это поле.

coder1coder committed:
Добавить возможность регистрации пользователя

+ изменена модель User, добавлена Role
+ добавлен класс, интерфейс репозитория пользователя
+ добавлен класс сервиса для работы с пользователем
+ добавлен метод регистрации в контроллер пользователей

// Непонятно кто регулирует назначение роли пользователя, поэтому DTO при регистрации не содержит это поле.

pingvin1308 commented on LessonMonitor/LessonMonitor.API/Infrastucture/Services/UserService.cs:
```diff
@@ -0,0 +1,26 @@
+
+        public Task<User> SignUpAsync(SignUpRequest request)
+        {
+            var mapper = new Mapper(new MapperConfiguration(x => x.CreateMap<SignUpRequest, User>()));
```
конфигурацию маппера лучше вынести отдельно и вместо Mapper использовать интерфейс IMapper, который будет прокинут в конструктор

pingvin1308 replied:
Также маппинг в доменную модель следует делать в контроллере, так как по хорошему бизнес логический слой ничего не должен знать об апи контрактах (про SignUpRequest)

pingvin1308 commented on LessonMonitor/LessonMonitor.API/Controllers/UsersController.cs:
```diff
@@ -134,5 +144,18 @@ public void GetModel([FromQuery] User user)
+        [HttpPost("[action]")]
+        public async Task<IActionResult> SignUp(SignUpRequest request)
+        {
+            return Ok(await _userService.SignUpAsync(request));
```
лучше сделать класс для возврата информации о юзере, который будет находиться в контрактном слое

pingvin1308 commented on LessonMonitor/LessonMonitor.API/Controllers/UsersController.cs:
```diff
@@ -4,50 +4,60 @@
+        private readonly IUserRepository _userRepository;
 
-        public UsersController(IUserRepository userRepository)
+        public UsersController(IUserService userService, IUserRepository userRepository)
```
Думаю можно убрать IUserRepository отсюда. Будем работать через сервисный класс

pingvin1308 commented on LessonMonitor/LessonMonitor.API/Application/DTOs/SignUpRequest.cs:
```diff
@@ -0,0 +1,26 @@
+
+        [Required(ErrorMessage = "Поле 'Возраст' обязательно для заполнения")]
+        [Range(1, 100, ErrorMessage = "Возраст пользователя должен быть от 1 до 100 ")]
+        public int Age { get; set; }
```
Думаю поле возраст должно быть необязательным)

coder1coder committed:
Правки по таску регистрации пользователя

+ убрал аттрибут Required у поля Age в DTO при регистрации пользователя
+ убрал интерфейс репозитория пользователя из конструктора UsersController
+ создал контрактный класс UserDTO для возврата информации о пользователе
+ вынес конфиги мапера, маппер через DI, маппинг из сервиса в контроллер

pingvin1308 commented on LessonMonitor/LessonMonitor.API/Application/DTOs/UserDTO.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using LessonMonitor.API.Domain;
+
+namespace LessonMonitor.API.Application.DTOs
```
Сделать директорию Models и убрать DTO

trulander commented:
Все рассортировано по своим директориям, но вот нихрена не понятно их назначение Domain Infrastucture Application. Чувствуется огромный разрыв между тем что мы проходим на занятиях и тем как тут реализовано. Как будтое я пропустил половину занятий, или Роман ведет секретные занятия где рассказывает все тонкости и нюансы продуктовой разработки:-D

Кстати ты не изменил таску в проекте который в основной репе, ты создал таску и пометил ее как выполненную только в своем репозитории.

pingvin1308 reviewed (commented):
> Все рассортировано по своим директориям, но вот нихрена не понятно их назначение Domain Infrastucture Application. Чувствуется огромный разрыв между тем что мы проходим на занятиях и тем как тут реализовано. Как будтое я пропустил половину занятий, или Роман ведет секретные занятия где рассказывает все тонкости и нюансы продуктовой разработки:-D
> 
> Кстати ты не изменил таску в проекте который в основной репе, ты создал таску и пометил ее как выполненную только в своем репозитории.

Так точно, введу секретные знания о том, почему так строится проект)

coder1coder commented:
> Кстати ты не изменил таску в проекте который в основной репе, ты создал таску и пометил ее как выполненную только в своем репозитории.

У меня нет доступа к проекту на этой репе. Я следую [инструкции](https://github.com/cleannetcode/LessonMonitor/wiki/%D0%9A%D0%B0%D0%BA-%D0%B2%D0%B7%D1%8F%D1%82%D1%8C%D1%81%D1%8F-%D0%B7%D0%B0-%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D1%83-%D1%81-%D0%B7%D0%B0%D0%B4%D0%B0%D1%87%D0%B0%D0%BC%D0%B8)
